ROI related tidying and changes; basic serial comms code

omarjamil opened the pull request:
Doing a poll request as the code changes are getting larger and I have changed a few things related to ROI both on the frontend and the server. I have tried to generalise the ROI creation on the server side so should be able to implement read-from-file functionality.

For the frontend, jquery was being used in a way that caused multiple firing onclick. Fixed that issue.

Added some basic arduino comms code; Also added an input box to create roi page to set the arduino serial port.

omarjamil commented:
@felicityguest It would be good to get this in before you do any more ROI work as this changes a lot of stuff related to that.

felicityguest commented on netclamp/filter/roi.py:
```diff
@@ -0,0 +1,49 @@
+    """
+    def __init__(self):
+        self.history = []
+        self.history_len = None 
```
Are `history` and `history_len` needed here? They're included in `RectangleROI` too.

felicityguest commented on netclamp/filter/roi.py:
```diff
@@ -0,0 +1,49 @@
+
+    def process_rois(self, image: np.ndarray, rois: list, threshold: float) -> list[float]:
+        """
+        Process the all the rois and apply thresholding
```
Nit: Don't need the first 'the' here.

felicityguest commented on netclamp/filter/roi.py:
```diff
@@ -0,0 +1,49 @@
+        """
+        Process the all the rois and apply thresholding
+        """
+        active_rois = list(np.zeros(len(rois),dtype=int))
```
Are we using a `list` instead of an `np.ndarray` here because of the Arduino?

omarjamil replied:
Yes, that's right. I think that will be easier to map than numpy objects.

felicityguest reviewed (approved):
I've made a few comments, but there don't seem to be any obvious issues. I'm happy for this PR to be merged and for us to fix any minor things in the next few weeks. We'll need to fix the ROI test at some point too.

omarjamil commented on app/api/camera.py:
```diff
@@ -87,8 +87,9 @@ async def capture_image():
+    print(frame.shape)
     for filter in Session.filters:
-        frame = filter.process(frame)
+        filter.process(frame)
```
This is processing all the ROIs, but not part of the experimental loop. Just for showing the changes in the graph on the frontend at the moment.

omarjamil commented on app/api/filter.py:
```diff
@@ -26,10 +26,8 @@ async def create_roi(roi: RegionOfInterestInput):
     mins = (roi.min_x, roi.min_y)
     maxs = (mins[0] + roi.delta_x, mins[1] + roi.delta_y)
```
Nope, can get rid of these.

omarjamil commented on app/templates/pages/create_regions.html:
```diff
@@ -111,7 +82,21 @@ <h4 style="margin-top: 8%; margin-bottom: 3%;">Camera Settings</h4>
+        Set
+      </button>
+     
+    </div>
```
This has made think we should integrate load ROI into the same page. Will make the changes.

omarjamil commented on app/templates/pages/create_regions.html:
```diff
@@ -2,33 +2,13 @@
+      <a href="https://github.com/UniExeterRSE/NetClamp" class="btn btn-info" role="button" target="_blank">Repo</a>
+      <a href="https://UniExeterRSE.github.io/NetClamp/" class="btn btn-info" role="button" target="_blank">Docs</a>
+      <a href="/docs" class="btn btn-info" role="button" target="_blank">API</a>
     </div>
```
I think we should remove load ROIs page as that is practically the same as create ROI with just one path to the ROI file. Might be best to integrate ROI file load path into create roi page.

omarjamil commented on netclamp/camera/hamamatsu.py:
```diff
@@ -73,17 +77,23 @@ def start_experiment(self, save_path: str, save_name: str):
-        ROI = RegionOfInterest()
-        ROI.create_rois_from_file()
+        roi = ROI()
+        cells = roi.create_rois_from_file("roi_file.json")
```
Not an indication of what format file to use. Just put something down here!

omarjamil commented on netclamp/filter/roi.py:
```diff
@@ -0,0 +1,49 @@
+
+class ROI(Filter):
+    """
+    Square region of interest.
```
Changed

omarjamil commented:
@felicityguest I have actioned your comments (thank you for those). Prompted by those I have done some further rationalisation and removed load ROI page and added the path loading to create_roi page as that might be simpler. Happyt discuss.

omarjamil commented:
Also adding arduino code that reads from serial port a list being sent from `controllers.py`. The list should be ones or zeros and the arduino activates digital pin 13 if the value is 1. The camera can be hooked up to pin 13 and this will trigger it.

felicityguest commented on app/templates/pages/load_regions.html:
```diff
@@ -2,33 +2,13 @@
 
```
Should we delete this page now? We're no longer using it.

omarjamil replied:
Fine by me. Please go ahead.

felicityguest commented on app/templates/pages/menu.html:
```diff
@@ -2,32 +2,15 @@
         <button
             class="btn btn-danger"
-            onclick="launch_camera('Webcam')"
+            onclick="launch_camera('Webcam', 'create')"
```
As there's only one setup page now, we could remove the additional argument for `launch_camera()`.

felicityguest commented on app/templates/config/webcam.html:
```diff
@@ -52,19 +52,19 @@ <h1 class="display-5 fw-bold" style="text-align: center">NetClamp</h1>
+    <!-- </div> -->
 </div>
 
 {% endblock %}
```
The `Repo`, `Docs` and `API` buttons also need to be changed in this file.

felicityguest commented on arduino/read_serial/read_serial.ino:
```diff
@@ -0,0 +1,44 @@
+void loop() {
+    checkForNewData();
+    if (newData == true) {
+        for (int n=0; n<=dataSize; n++)
```
Should it be `n<dataSize`?

omarjamil replied:
Yes. @felicityguest please make that change. I have tested it, but not with proper regions data.

felicityguest reviewed (approved):
Again, I've made a few comments but there are no obvious blockers to this PR. I presume you've tested the Arduino code, so my question regarding the `for` loop in `loop()` can probably be ignored.

felicityguest commented:
I'll remove the `load_regions` page, amend `launch_camera()` and update the buttons (`Repo`, `Docs`, `API`) shortly.

omarjamil commented:
@felicityguest Once you are happy with it and implemented the changes, please go ahead and merge this.
